### LIST OF SUBSEQUENCES

1. example ->"abc"
1. base case : -> empty or single length, take empty string so that single length can be managed by empty string itself, return array with 1 size 
2.  Induction Hypo : for all length k, we find subsequences
2. Induction Step: Assume funcn will work fine for n-1, and that smaller string output will give the right output of the bigger string
3. subsequences have a or not, b or not, c or not
4. so total subsequences = 2^n
5. if we have output for bc ie "", b, c, bc;; then for abc new output will have either a or not a , building total 2* 4 characters

In [3]:
def subs(s):
    if len(s)==0:
        output=[] ## array made-2
        output.append("") ## appended asn returned to c-3
        return output
    
    smallerString=s[1:] ## rchd till 0 -1
    smallerOutput = subs(smallerString)
    
    output=[] ## another array for c-4
    for sub in smallerOutput:
        output.append(sub) ## "" built as at ease
        
    for sub in smallerOutput:
        subs_with_zero_char = s[0]+sub ## c+"" built
        output.append(subs_with_zero_char)
        
    return output

In [4]:
subs("abcd")

['',
 'd',
 'c',
 'cd',
 'b',
 'bd',
 'bc',
 'bcd',
 'a',
 'ad',
 'ac',
 'acd',
 'ab',
 'abd',
 'abc',
 'abcd']

### RETURN KEYPAD COMBINATIONS
1. number is given ie 24
2. 2->abc, 4->ghi
3. possible strings: ag,ah,ai,bg,bh,bi,cg,ch,ci
4. possible no. of outputs = 3* 3
5. assuming 1 and 0 arent part of this, 1 corresponds to special characters, 0 to space
6. 2 corresponds to "a","b","c"

1. 234->23->2->empty string
2. 2-'abc';2-'a,b,c';23-'ad,ae,af,bd,be,bf,cd,ce,cf'
3. smallerind = input/10
4. remaining digits-input %10
5. smalleroutput=getsub(smallerind)
6. output=[]

In [5]:
def getString(d):
    if d ==2:
        return "abc"
    
    if d==3:
        return "def"
    
    if d==4:
        return "ghi"
    
    if d ==5:
        return "jkl"
    
    if d==6:
        return "mno"
    
    if d==7:
        return "pqrs"
    
    if d ==8:
        return "tuv"
    
    if d==9:
        return "wxyz"
    
    
    return " "

In [8]:
def keypad(n):
    if n==0:
        output=[]
        output.append("")
        return output
    
    smallerNumber=n//10
    lastDigit=n%10
    
    smallerOutput=keypad(smallerNumber)
    optionsForLastDigit = getString(lastDigit)
    
    output=[]
    
    for s in smallerOutput :
        for c in optionsForLastDigit:
            option = s+c
            output.append(option)
    return output


In [11]:
keypad(23)

['ad', 'ae', 'af', 'bd', 'be', 'bf', 'cd', 'ce', 'cf']

In [10]:
output=keypad(237)
len(output), output

(36,
 ['adp',
  'adq',
  'adr',
  'ads',
  'aep',
  'aeq',
  'aer',
  'aes',
  'afp',
  'afq',
  'afr',
  'afs',
  'bdp',
  'bdq',
  'bdr',
  'bds',
  'bep',
  'beq',
  'ber',
  'bes',
  'bfp',
  'bfq',
  'bfr',
  'bfs',
  'cdp',
  'cdq',
  'cdr',
  'cds',
  'cep',
  'ceq',
  'cer',
  'ces',
  'cfp',
  'cfq',
  'cfr',
  'cfs'])

#### PRINT OUTPUT INSTEAD OF RETURNING

In [ ]:
def fact(n):
    if n==0:
        return 1
    smallerOutput=fact(n-1)
    
    print(n*smallerOutput) ## induction is gone coz returning hasnt been done for thte smaller input

In [12]:
def fact(n):
    if n==0:
        return 1
    smallOutput = fact(n-1)
    output=n*smallOutput
    print(output)
    return output

In [13]:
fact(5)

1
2
6
24
120


120

In [14]:
def factHelper(n):
    if n==0:
        return 1
    smallOutput = factHelper(n-1)
    output=n*smallOutput
    return output

In [15]:
def fact(n):
    output=factHelper(n)
    print(output)

In [16]:
fact(5)

120


##### still not printing from the actual funcn

1. building the ans from base case to upwards

2. Build the answer while gng down, calling for n -? ->  call  for n-1 and tell ans is n, and keep passing on thee answer downwards

3. rch the base case and then print at the end


In [17]:
def printFact(n,ans):
    if n == 0:
        print(ans)
        return
    ans=ans*n
    printFact(n-1,ans)

In [19]:
printFact(5,1)

120


1. calling on 5 -> will call 4 but before that ans will be 1*5= 5
2. CALLS ARE AS FOLLOW: 5,1->4,5->3,20->2,60->1,120->0,120

#### MINIMUM OF ARRAY  - PRINNT AND NOT RETURN

1. Given a list we need to return the minimum list
2. BASE CASE - empty string, will return a +ve, or 1 single ele and returns itself
3. N-1 -> return minimum and then check for min in nth ele and that returned from n-1

In [ ]:
def minList(1):
    if len(l)==1:
        return l[0]
    minSmallArray = minList(l[1:])
    overallMin=min(minSmallArray,l[0])
    return overallMin

In [20]:
def printMin(l,minsofar=100000):
    if len(l)==0:
        print(minsofar)
        return
    newMin=min(minsofar,l[0])
    printMin(l[1:],newMin)

[3,4,2,-1,0,9,-2,5]
1. 10000 and 3, 3 will be passed, 3 again, 2,-1,-1,-1,-2,-2

In [22]:
printMin([1,2,3,-4,-1,0,5,-6])

-6


### PRINT ALL SUBSEQUENCES
1. abc was completing its work after bc in case of RETURN
2. in case of print we want, abc to do its work and then go to bc
3. 2 CASES:
    1. abc -> bc,"" and bc,"a"
    2. bc,""->c,""; c,"b" and bc,"a"-> c,"a" and c,"ab"
4. decision to make and run both sides on abc; therefore we will have 2 recursive calls
5. string; input, output
6. call on smaller part of input once excluding the first charac and then other time on including the first charac

###### basically u would be printing the bases cases

In [23]:
def printSubs(s,o):
    if len(s)==0:
        print(o)
        return
    ## not including 0 char
    printSubs(s[1:],o)      ## this part will be completed first 
    
    ## including 0 char
    newOutput=o+s[0]
    printSubs(s[1:],newOutput)

In [24]:
printSubs("abc","")


c
b
bc
a
ac
ab
abc


In [25]:
string=input()
printSubs(string,"")

abc

c
b
bc
a
ac
ab
abc


In [20]:
def subs(l):
    if l==[]:
        return [[]]
    
    x=subs(l[1:])
    
    
    return x+[l[0]+y for y in x]
x=[int(ele) for ele in input().split()]
for i in x:
    for j in i:
        print(j,end=" ")
    print()
    


1 2 3 4


TypeError: 'int' object is not iterable

### PRINT KEYPAD COMBINATIONS

1. HEAVY RECURSIVE CALLS and printing the base cases, 
2. TIME COMPLEXITY IS SIMILAR
3. In prev solution , we were doing lots of work in getting the huge array

1. if input==0:
        1.print(outputsofar)
        2. return
2. s=GetOptions(LastDigit):
    1. s looped

In [26]:
def getString(d):
    if d ==2:
        return "abc"
    
    if d==3:
        return "def"
    
    if d==4:
        return "ghi"
    
    if d ==5:
        return "jkl"
    
    if d==6:
        return "mno"
    
    if d==7:
        return "pqrs"
    
    if d ==8:
        return "tuv"
    
    if d==9:
        return "wxyz"
    
    
    return " "

In [29]:
def printkeypad(i,o):
    if i==0:
        print(o)
        return
    
    smallnumber=i//10
    lastdigit=i%10
    
    optionsforLastDigit=getString(lastdigit)
    for c in optionsforLastDigit:
        newOutput=c+o
        printkeypad(smallnumber,newOutput)

In [31]:
printkeypad(23,"")

ad
bd
cd
ae
be
ce
af
bf
cf


### RETURN SUBSET SUM TO K
1. 3 2 5 1 4 6, K = 8 ; ANS BEING  3 5; 3 1 4; 2 5 1 ; 2 6
2. 3 being included in some cases, 3 not being included in other cases
3. 3 included, take rest of the input array and get the inputs which sum to 5
4. 3 not included, pass rest of the array and give array sum to 8 



In [3]:
## to get in the same order as inputted in the string
def printSubsetToKHelper(arr,k,output):
    if len(arr)==0:
        if k==0:
            for ele in output:
                print(ele,end=' ')
            print()
            return
        else:
            return
        ##excluding the first item
    printSubsetToKHelper(arr[1:],k,output)
        
    output1=[ele for ele in output]
    output1.append(arr[0])
        
        ##including the first ele
    printSubsetToKHelper(arr[1:],k-arr[0],output1)
def printSubsetToK(arr,k):
    output=[]
    printSubsetToKHelper(arr,k,output)

n=int(input())
arr=[int(ele) for ele in input().split()]
k=int(input())

printSubsetToK(arr,k)

9
5 12 3 17 1 18 15 3 17 
6
3 3 
5 1 


### PRINT SUBSET SUM TO K

In [38]:
def printSubsetToK(arr,n,v,sum):
    if(sum==0):
        for value in v:
            print(value,end=" ")
        print()
        return
    if(n==0):
        return
    printSubsetToK(arr,n-1,v,sum)
    
    v1=[]+v
    v1.append(arr[n-1])
    printSubsetToK(arr,n-1,v1,sum-arr[n-1])
    
n=int(input())
arr=[int(ele) for ele in input().split( )]
k=int(input())

printSubsetToK(arr,n,[],k)

9
5 12 3 4 1 3 7 8 9
6
1 5 
3 3 


### PRINT SUBSET SUM TO K

In [1]:
def printSubset(arr,k,lst):
    n=len(arr)
    if n==0:
        return
    if n==1:
        if arr[0]==k:
            print(k,*lst)
            return
        
    printSubset(arr[:-1],k,lst)
    printSubset(arr[:-1],k-arr[-1],[arr[-1]]+lst)
    if arr[-1]==k:
        print(k,*lst)
        
def printSubsetMain(arr,k):
    printSubset(arr,k,[])
    
n=int(input())
arr=[int(ele) for ele in list(input().split())]
k=int(input())
printSubsetMain(arr,k)

9
5 12 3 4 1 3 7 8 9
6
5 1
3 3


In [31]:
def printAllSubsetsRec(arr,n,v,k):
    if (k==0):
        for value in v:
            print(value,end=" ")
        print()
        return
    if (n==0):
        return
    
    printAllSubsetsRec(arr,n-1,v,k)
    v1=[]
    v1.append(arr[n-1])
    v1+=v
    printAllSubsetsRec(arr,n-1,v1,k-arr[n-1])

def printAllSubsets(arr,n,k):
    v=[]
    printAllSubsetsRec(arr,n,v,k)
    
n=int(input())
arr=[int (ele) for ele in input().split()]
k=int(input())
printAllSubsets(arr,n,k)

5
1 2 3 3 4 5
6
1 2 3 
1 2 3 
3 3 
2 4 


### RETURN PERMUTATIONS OF A STRING

In [6]:
## didnt get all the outputs 
def permutations(string):
    n=len(string)
    
    if n==0:
        return
    output=[]
    for i in range (n):
        start=string[i]
        
        remaining=string[:i]+string[i+1:]
        permutations(remaining)
        output.append(start+remaining)
        
        next1=remaining[::-1]
        permutations(next1)
        output.append(start+next1)
        
    return output

string = input()
ans = permutations(string)
for s in ans:
    print(s)


bcdlp
bcdlp
bpldc
cbdlp
cpldb
dbclp
dplcb
lbcdp
lpdcb
pbcdl
pldcb


In [4]:
def permutations(string):
    """
    Create all permutations of a string with non-repeating characters
    """
    permutation_list = []
    if len(string) == 1:
        return [string]
    else:
        for char in string:
            [permutation_list.append(char + a) for a in permutations(string.replace(char, "", 1))]
    return permutation_list
string=input()
ans=permutations(string)
print(*ans)
print(len(ans))

bcdlp
bcdlp bcdpl bcldp bclpd bcpdl bcpld bdclp bdcpl bdlcp bdlpc bdpcl bdplc blcdp blcpd bldcp bldpc blpcd blpdc bpcdl bpcld bpdcl bpdlc bplcd bpldc cbdlp cbdpl cbldp cblpd cbpdl cbpld cdblp cdbpl cdlbp cdlpb cdpbl cdplb clbdp clbpd cldbp cldpb clpbd clpdb cpbdl cpbld cpdbl cpdlb cplbd cpldb dbclp dbcpl dblcp dblpc dbpcl dbplc dcblp dcbpl dclbp dclpb dcpbl dcplb dlbcp dlbpc dlcbp dlcpb dlpbc dlpcb dpbcl dpblc dpcbl dpclb dplbc dplcb lbcdp lbcpd lbdcp lbdpc lbpcd lbpdc lcbdp lcbpd lcdbp lcdpb lcpbd lcpdb ldbcp ldbpc ldcbp ldcpb ldpbc ldpcb lpbcd lpbdc lpcbd lpcdb lpdbc lpdcb pbcdl pbcld pbdcl pbdlc pblcd pbldc pcbdl pcbld pcdbl pcdlb pclbd pcldb pdbcl pdblc pdcbl pdclb pdlbc pdlcb plbcd plbdc plcbd plcdb pldbc pldcb
120


In [13]:
from itertools import permutations

def totalpermutations(str):
    
    permList = permutations(str)
    for perm in list(permList):
        print(''.join(perm))


string = input()
ans = totalpermutations(string)


abc
abc
acb
bac
bca
cab
cba


### PRINT PERMUTATIONS OF A STRING

In [33]:
def printPermutations(string,output):
    if len(string)==0:
        print(output)
        return
    for i in range(len(string)):
        printPermutations(string[0:i]+string[i+1:],output+string[i])
string=input()
printPermutations(string,'')

abc
abc
acb
bac
bca
cab
cba


### RETURN SUBSETS OF AN ARRAY

In [25]:
def returnSubsets(a,n):
    if n==0:
        return [[]]
    smallout=returnSubsets(a[1:],n-1)
    including=[]
    
    for i in smallout:
        temp=[a[0]]
        for j in i:
            temp.append(j)
        including.append(temp)
    out=[]
    for i in including:
        out.append(i)
    for j in smallout:
        out.append(j)
    return out

n=int(input())
a=[int(i) for i in input().split()]
for i in returnSubsets(a,n):
    for j in i:
        print(j,end=' ')
    print()

3
12 32 14
12 32 14 
12 32 
12 14 
12 
32 14 
32 
14 



### PRINT SUBSETS OF AN ARRAY 

In [21]:
def printSubsnH(s,o):
    if len(s)==0:
        print(o)
        return
    ## not including 0 char
    printSubsn(s[1:],o)      ## this part will be completed first 
    
    ## including 0 char
    newOutput=str(o)+" "+ str(s[0])
    printSubsn(s[1:],newOutput)
def printSubsn(arr,out):
    printSubsnH(arr,out)
arr=[int(ele) for ele in input().split()]
printSubsn(arr,"")

5 10 12

 12
 10
 10 12
 5
 5 12
 5 10
 5 10 12


#### PRINT SUBSETS OF AN ARRAY

In [29]:
def printSubsets(arr,si,output):
    if si>=len(arr):
        for i in output:
            print(i,end=' ')
        print()
        return
    newOutput=[] ### empty array-1
    for i in output: 
        newOutput.append(i) ##appended for each i-5
    newOutput.append(arr[si]) ## added the first ele in it-2
    printSubsets(arr,si+1,newOutput) ## while using the first ele-3
    printSubsets(arr,si+1,output) ## while not using the first ele -4
    
n=int(input())
arr=[int(ele) for ele in input().split(" ")]
printSubsets(arr,0,[]) 

3
12 20 15
12 20 15 
12 20 
12 15 
12 
20 15 
20 
15 

